docs/ocis/adr/0016-files-metadata.md

```diff
@@ -0,0 +1,84 @@
+
+## Decision Outcome
+
+Chosen option: "Extended File Attributes", because we guarantee the consistency of data and have arbitrary simple storage mechanism.
+
+### Positive Consequences
+
+* Metadata is always attached to the file itself
+* We can store arbitrary key/values
+* No external dependencies are needed
+
+### Negative consequences
+
+* The storage inside extended file attributes has limits
+* Changes to extended attributes are not atomic and need file locks
+
+## Pros and Cons of the Options <!-- optional -->
+
+### Database or Key-Value Store
+
+Use a Database or an external key/value store to persist metadata.
+
+* Good, because it scales well
+* Good, because databases provide efficient lookup mechanisms
+* Bad, because the file content and the metadata could run out of sync
+* Bad, because a storage backup doesn't cover the file metadata
+
+### Extended File Attributes
+
+Extended File Attributes allow to store arbitrary properties. There are 4 namespaces `user`, `system`, `trusted` and `security`. We can safely use the `user` namespace. An example attribute name would be `user.ocis.owner.id`. The linux kernel has length limits on attribute names and values.
+
+From Wikipedia on [Extended file attributes](https://en.wikipedia.org/wiki/Extended_file_attributes#Linux):
+
+> The Linux kernel allows extended attribute to have names of up to 255 bytes and values of up to 64 KiB,[14] as do XFS and ReiserFS, but ext2/3/4 and btrfs impose much smaller limits, requiring all the attributes (names and values) of one file to fit in one “filesystem block” (usually 4 KiB). Per POSIX.1e,[citation needed] the names are required to start with one of security, system, trusted, and user plus a period. This defines the four namespaces of extended attributes.
+
+* Good, because metadata is stored in the filesystem
+* Good, because consistency is easy to maintain
+* Good, because the data is attached to the file and survives file operations like copy and move
+* Good, because a storage backup also covers the file metadata
+* Bad, because we could hit the filesystem limit
+* Bad, because changes to extended attributes are not atomic
+
+### Metadata File
+
+We could store metadata in a metadata file next to the file content which has a structured content format like .json, .yaml or .toml. That would give us more space to store bigger amounts of metadata.
+
+* Good, because there are no size limits
+* Good, because there is more freedom to the content format
+* Good, because a storage backup also covers the file metadata
+* Bad, because it doubles the amount of read / write operations
+* Bad, because it needs additional measures against concurrent overwriting changes
+
+### Link metadata with an id in the extended attributes
+
+To link metadata to file content a single extended attribute with a file id (unique per storage space) is sufficient. This would also allow putting metadata in better suited storage systems like SQLite or a key value store.
```

Comment:
Does that mean that the file id is the name of the file that contains the metadata?

Reply:
Probably you do not want to specify that here on purpose... Ok.

Reply:
that really is up to the implementation. in decomposedfs we could write json files instead of symlinks which then contain all necessary data. but then you can no longer navigate through the fs on the cli. a tradeoff I would be willing to make if it makes the fs simpler or more atomic. for other storage drivers the implementaition certainly looks different.